_ramblings/2018-07-11-building-a-shell-part-1.md

```diff
@@ -0,0 +1,203 @@
+functional shell. We'll use Ruby for this, because it'll make things very easy,
+as you'll soon see.
+
+### Our Overly Simplistic Ruby Shell
+
+The acronym REPL gives us a starting point for designing our shell, with a
+wishlist of functions to implement. We can get away without defining `print` or
+`loop` since evaluation will print itself, and [`loop`][loop] is built into
+Ruby.
+
+- `read`
+- `evaluate`
+
+We'll start by implementing the `read` function. This function will block until
+an entire line is read using [`gets`][gets]. A much more powerful shell might
+read in pieces to update syntax highlighting as you type, or provide various
+kinds of completion. However, we're not building anything that fancy at the
+moment.
+
+```rb
+# Print a prompt, then read a line from STDIN.
+def read
+  print('$ ')
+
+  # Try to read a line, exiting if there's no more lines.
+  line = STDIN.gets
+  if line
+    line.chomp.split(' ')
+  else
+    exit
+  end
+end
+```
+
+The following implementation of `evaluate` is a copout, because we just pass
+the input through to Ruby's implementation of [`system`][system] which calls
+`/bin/sh` under the hood. It does the correct thing however, and is enough to
+get us off the ground running.
+
+```rb
+# Run the given command in a subprocess.
+def evaluate(argv)
+  success = system(*argv)
+  unless success
+    puts("unknown command '#{argv[0]}'")
+  end
+end
+```
+
+Now that we've defined both `read` and `evaluate`, the full blown REPL is a
+simple one liner.
+
+```rb
+# The glorious REPL itself.
+loop { evaluate(read) }
+```
+
+Running this tiny Ruby program will give you a semi-functional shell. Take a
+moment and try it out for yourself! There are however, a few critical features
+missing, for example you'll notice `cd` doesn't work. This is where builtins
+come into play.
+
+### Extending Our Shell (a little)
+
+Builtins are commands that the shell executes directly, without calling out
+to a subprocess. This is exactly what we need for `cd`, since we don't want to
+change the working directory of a sub-process, we want to update the shell's
+working directory itself.
+
+We'll define a hash of `'command' => lambda { |args| ... }` to serve as a
+lookup table for the builtins.
+
+```rb
+# Commands to be executed directly by this shell.
+BUILTINS = {}
+```
+
+Next we need to update our `evaluate` function to handle our special builtin
+commands.
+
+```rb
+# Run the given command in a subprocess.
+def evaluate(argv)
+  if BUILTINS[argv[0]]
+    BUILTINS[argv[0]].call(argv[1..-1])
+  else
+    success = system(*argv)
+    unless success
+      puts("unknown command '#{argv[0]}'")
+    end
+  end
+end
+```
+
+Finally, we can start defining the needed builtins. We'll start with the two
+most needed commands, `cd` and `exit`, however there are many others which
+could be defined at this point. Mainly `.` (or `source`), `alias` and `exec`.
+I'll leave those implementations as exercises for the reader.
```

Comment:
As a reader, the turn of phrase makes me feel patronised.

Reply:
I'll chalk this up to style. I personally love reading things like this. Gives me that extra umph to actually go try what's being suggested. I do understand it *is* patronizing, just hopefully not in a way that's overly offensive.

Reply:
Is inoffensively patronizing a thing?

Reply:
My favorite professors had something like that going on. I guess I'm trying to mirror that a bit.